src/hostapi/pulseaudio/pa_linux_pulseaudio.c

```diff
@@ -1311,3 +1329,86 @@ double GetStreamCpuLoad( PaStream * s )
 
     return PaUtil_GetCpuLoad( &stream->cpuLoadMeasurer );
 }
+
+/** Extensions */
+static void RenameStreamCb(pa_stream *s, int success, void *userdata)
+{
+    /* Currently does nothing but signal the caller. */
+    PaPulseAudio_Stream *l_ptrStream = (PaPulseAudio_Stream *) userdata;
+    pa_threaded_mainloop_signal( l_ptrStream->mainloop,
+                                 0 );
+}
+
+PaError PaPulseAudio_RenameSource( PaStream *s, const char *streamName )
+{
+    PaPulseAudio_Stream *stream = (PaPulseAudio_Stream *) s;
+    PaError result = paNoError;
+    pa_operation *op = NULL;
+
+    if ( stream->inStream == NULL )
+    {
+        return paInvalidDevice;
+    }
+
+    /* Reallocate stream name in memory. */
+    pa_threaded_mainloop_lock( stream->mainloop );
+    char *newStreamName = (char*)PaUtil_AllocateMemory(strnlen(streamName, PAPULSEAUDIO_MAX_DEVICENAME) + 1);
+    if ( !newStreamName )
+    {
+        pa_threaded_mainloop_unlock( stream->mainloop );
+        return paInsufficientMemory;
+    }
+    memset(newStreamName, 0, strnlen(streamName, PAPULSEAUDIO_MAX_DEVICENAME) + 1);
+    strncpy(newStreamName, streamName, strnlen(streamName, PAPULSEAUDIO_MAX_DEVICENAME));
+
+    PaUtil_FreeMemory( stream->sourceStreamName );
+    stream->sourceStreamName = newStreamName;
+
+    op = pa_stream_set_name( stream->inStream, streamName, RenameStreamCb, stream );
+    pa_threaded_mainloop_unlock( stream->mainloop );
+
+    /* Wait for completion. */
+    while (pa_operation_get_state( op ) == PA_OPERATION_RUNNING)
+    {
+        pa_threaded_mainloop_wait( stream->mainloop );
+    }
+
+    return result;
+}
+
+PaError PaPulseAudio_RenameSink( PaStream *s, const char *streamName )
+{
+    PaPulseAudio_Stream *stream = (PaPulseAudio_Stream *) s;
+    PaError result = paNoError;
+    pa_operation *op = NULL;
+    
+    if ( stream->outStream == NULL )
+    {
+        return paInvalidDevice;
+    }
+
+    /* Reallocate stream name in memory. */
+    pa_threaded_mainloop_lock( stream->mainloop );
```

Comment:
we need to check if we actually have a mainloop. I think this lock should only guard the pulse calls.

Reply:
The main loop is created during PortAudio initialization, correct? (PaPulseAudio_Initialize/PaPulseAudio_New)

I also remember having to move around the lock/unlock calls a bit because I was running into crash issues related to allocating/deallocating the stream names. Should we possibly be using pa_xalloc() and similar instead for those?

Reply:
Yes.. as mainloop runs and if you don't lock it you will suffer.. and it there is no mainloop nothing works so.. for stream naming before init we have to come to another solution.

Reply:
That seems like a different issue from what @daschuer was mentioning. For this, there shouldn't be an issue since streams can't be created/started without PortAudio having been initialized via Pa_Initialize().